### ISSUE TITLE
Beartype error with Python 3.6

### ISSUE THREAD

**Issue by daxpryce** [beartype/beartype#100]
@leycec beartype 0.10.0 seems to have an issue with Python 3.6 - and/or we're using it wrong, which is possible.

https://github.com/microsoft/graspologic/runs/5133579368?check_suite_focus=true#step:5:17

```
tests/partition/test_leiden.py:11: in <module>
    from beartype.roar import BeartypeCallHintPepParamException
E   ImportError: cannot import name 'BeartypeCallHintPepParamException'
```

Reproduced it locally with:
```bash
dax@cloudship graspologic: $ python --version
Python 3.6.15
dax@cloudship graspologic: $ python -m venv venv
dax@cloudship graspologic: $ source venv/bin/activate
(venv) dax@cloudship graspologic: $ pip install beartype
Collecting beartype
  Downloading https://files.pythonhosted.org/packages/75/24/01fe5e88cc237293c24fb6f8a167b805df89fecfcfa47ec0e6f9e4eda4cd/beartype-0.10.0-py3-none-any.whl (571kB)
    100% |████████████████████████████████| 573kB 10.7MB/s
Installing collected packages: beartype
Successfully installed beartype-0.10.0
You are using pip version 18.1, however version 21.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
(venv) dax@cloudship graspologic: $ python -c "from beartype.roar import BeartypeCallHintPepParamException"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: cannot import name 'BeartypeCallHintPepParamException'
(venv) dax@cloudship graspologic: $ deactivate
dax@cloudship graspologic: $ pyenv local 3.9.10
dax@cloudship graspologic: $ rm -rf venv
dax@cloudship graspologic: $ python -m venv venv
dax@cloudship graspologic: $ source venv/bin/activate
(venv) dax@cloudship graspologic: $ pip install beartype
Collecting beartype
  Using cached beartype-0.10.0-py3-none-any.whl (571 kB)
Installing collected packages: beartype
Successfully installed beartype-0.10.0
WARNING: You are using pip version 21.2.4; however, version 22.0.3 is available.
You should consider upgrading via the '/Users/dax/source/graspologic/venv/bin/python -m pip install --upgrade pip' command.
(venv) dax@cloudship graspologic: $ python --version
Python 3.9.10
(venv) dax@cloudship graspologic: $ python -c "from beartype.roar import BeartypeCallHintPepParamException"
```

**Comment by leycec** [comment]
**Ah-ha!** I can actually field this one successfully for the greater good of Microsoft. `beartype 0.10.0` deprecated three badly named exception types:

* `beartype.roar.BeartypeCallHintPepException`, deprecated by `beartype.roar.BeartypeCallHintViolation`.
* `beartype.roar.BeartypeCallHintPepParamException`, deprecated by `beartype.roar.BeartypeCallHintParamViolation`.
* `beartype.roar.BeartypeCallHintPepReturnException`, deprecated by `beartype.roar.BeartypeCallHintReturnViolation`.

For sanity, `beartype` uses [PEP 562-compliant module-scoped `__getattr__()` dunder functions](https://www.python.org/dev/peps/pep-0562) to deprecate attributes. Of course, Python 3.6 fails to support PEP 562 and thus ignores those functions. This is why we're having an awkward moment. :sweat_smile: 

Thankfully for all, Python 3.6 officially hit its End of Life (EOL) two months ago and now *kinda* constitutes a compelling security risk. With humble apologies and profuse grovelling, this means @beartype will be dropping Python 3.6 support shortly. I was considering ripping it out entirely from `beartype 0.10.0` – but one thing led to another, the cats were angrily hissing at one another, and here we are... *still having an awkward moment.*

## tl;dr

Please consider dropping Python 3.6 support to avoid further balding on my part. Alternately, please globally substitute `beartype.roar.BeartypeCallHintPepParamException` with `beartype.roar.BeartypeCallHintParamViolation` and `beartype.roar.BeartypeCallHintPepReturnException` with `beartype.roar.BeartypeCallHintReturnViolation`.

Microsoft: "We are sorry."

**Comment by daxpryce** [comment]
> Please consider dropping Python 3.6 support to avoid further balding on my part.

Further balding on *our* part, Cecil!  I think it's time to at least open the conversation, if not pull the trigger.  Onward and upward, right?

**Comment by leycec** [comment]
![what would flipper do?](https://c.tenor.com/ji7t2l9S0uEAAAAd/delfin-dolphin.gif)
<sup>**Surfer:** `graspologic`</sup>
<sup>**Dolphin:** *Python 3.6*</sup>

**Comment by daxpryce** [comment]
Alright then!  So from our perspective, we're ditching 3.6 - which does save us some time.  We already had other dependencies decide it was time to drop 3.6 support (which I'm not judging them for) in a [_patch_ release](https://semver.org/) (which I **am** judging them for), AND we are just about to have our first major release - so now is the optimal time to rofl-stomp our contracts, if any time is.

Thanks for the response @leycec - we appreciate the work you do, and your erstwhile responsiveness sets a bar that's hard to meet for any open source project!
